### ISSUE TITLE
[selectors] `/idref()/` combinator to follow element references for arbitrary IDREF attributes

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#10970]
In the past we have been discussing a `/for/` combinator to go from a label to the input it references (cannot find the issue now). However, HTML is rife with several IDREF attributes. Here is a small subset:

- `for`, in `<label>` and `<output>`
- `list` in `<input>`
- A host of ARIA attributes (e.g. `aria-describedby`, `aria-labelledby`, `aria-activedescendant`, `aria-controls`, `aria-details`, `aria-flowto`, `aria-owns` etc.)
- Popovers (`popovertarget`)
- Invokers (`invoketarget`)
- [`anchor`](https://github.com/whatwg/html/pull/9144)
- Plus, web component authors can always define their own, custom IDREF attributes

It seems clear that the path forwards for future combinators is `'/' <ident> '/'`. But that also opens the path for having *functional* combinators. What if we define a combinator that takes any attribute as a parameter and treats it as an IDREF?
Then `/for/` becomes:

```css
label /idref(for)/ input {
	/* ... */
}
```

but also things like this are possible:

```css
input.country /idref(list)/ datalist {
	/* ... */
}
```

Or even:

```css
my-tooltip /idref(for)/ * {
	cursor: help;
}
```

**Comment by bradkemper** [comment]
A pseudo element would be less useful, because you wouldn’t be able to select descendants. This, something like this is currently illegal:

```
label::ref(for) span {}

**Comment by bradkemper** [comment]
> In the past we have been discussing a /for/ combinator to go from a label to the input it references (cannot find the issue now). 

Maybe this one?:

https://github.com/w3c/csswg-drafts/issues/397

**Comment by bradkemper** [comment]
I like this idea as far as it goes. The issue I posted above is about going the other direction, i.e. selecting the label from the input. But that’s a separate (important) problem. @LeaVerou’s idea here does sound very useful to me.

**Comment by LeaVerou** [comment]
> Wouldn’t something like `label::ref(for)` be more compatible and simple?

The fact that we have contorted pseudo-elements to facilitate encapsulation in some shadow DOM use cases doesn't mean they are an appropriate replacement for combinators. We should avoid pseudo-elements for targeting entire elements in the light DOM. That’s _precisely_ what combinators are for.

(and IMO we should even ditch pseudo-elements for the shadow DOM use cases, see #7922 and #10939)

**Comment by therealglazou** [comment]
Hellooooo Lea :-)

Selection of ID/IDREF agnosticly from the attributes names is an old dream of mine. Long ago, SGML had "paired elements", linked through that mechanism and their styling required such a selecting hint. For those who were around back in the nineties, that was a topic I discussed during the famous Clamart CSS WG meeting.

I think the proposed combinator is not enough, let me explain... The whole thing is, at its roots, about catching a given value from a named attribute on an arbitrary element and finding another arbitrary element catching another named attribute of same value. The common denominator here is the value, the attribute names are only carriers.
The proposed combinator, as a functional thing, specifies only one of these two attributes where it should propose two, the target attribute and the source one, the latter defaulting to "id".

So two choices:

1. elaborate a mechanism allowing to "store" and "reuse" an attribute value (would be super powerful but CSS isn't a programming language, right ? :-)
2. make your functional idref combinator dual-valued with an "id" default if the second parameter is missing

**Comment by LeaVerou** [comment]
@therealglazou I do think there is value in having an even more general combinator that lets you customize the attribute too, but for the sake of simplicity, I thought this is a good first step. Also, I couldn't find a good name/syntax for it that wouldn't complicate the common IDREF case. If you can, that would be great!

**Comment by therealglazou** [comment]
@LeaVerou I gave second thoughts to this PR. The problem with this combinator is that it has impact on attribute existence above and below the combinator itself. And if the combinator is "generic" enough to allow arbitrary source attributes, the selection algorithm does not know what target attribute has to be present before the combinator is hit in the reverse processing of the chain of selectors. I am therefore not sure any more that a combinator is the optimal solution here. I'm not saying we have much better options, I'm only saying it's going to impact the selection algo in a way that I don't really like... Not sure I'm fully understandable, let me know if it's not clear enough.

**Comment by LeaVerou** [comment]
Yeah, I'm afraid I don't understand what you mean at all 😕

**Comment by Psychpsyo** [comment]
> A pseudo element would be less useful, because you wouldn’t be able to select descendants. This, something like this is currently illegal:
> 
> ```
> label::ref(for) span {}
> ```

Pseudo-elements with [internal structure](https://www.w3.org/TR/selectors-4/#pseudo-element-structure) are allowed to have child- and descendant combinators after them.
(The places I know this is used in are [view-transitions-1](https://www.w3.org/TR/css-view-transitions-1/) and [css-forms-1](https://drafts.csswg.org/css-forms-1/))

**Comment by lcchueri** [comment]
What happens if `e1 /idref(for)/ e2` references a parent e1 descends ?

**Comment by Psychpsyo** [comment]
> What happens if `e1 /idref(for)/ e2` references a parent e1 descends ?

Going by [the current spec](https://drafts.csswg.org/selectors-5/#idref-combinators) for this:
1. It'd be `e1 /for/ e2`.
2. Parent-child relationships do not affect this at all. If `e1` has a `for` attribute and the element that that points to matches `e2`, then `e1 /for/ e2` matches the element that matched `e2`.
(IF the `for` attribute is an IDREF. So e1 would have to be a `label`.)

**Comment by LeaVerou** [comment]
> > What happens if `e1 /idref(for)/ e2` references a parent e1 descends ?
> 
> Going by [the current spec](https://drafts.csswg.org/selectors-5/#idref-combinators) for this:

Oh interesting, I hadn't noticed this was in the draft. I just opened #12451 to get a formal resolution that we'll use the slash syntax for all new combinators, which means that needs to become less generic.

But yes, to answer the initial question, `/idref(for)/` is unaffected by other tree relationships. So, for labels, to get all labels one would need something like `label /idref(for)/ *, label:not([for]) :is(input, select, textarea, ...)`, which is awkward but I think it's worth the flexibility of not tying this to HTML form semantics. We could later add a shortcut that is specific to form elements (and/or a pseudo-class to target all form associated elements, including custom elements).

Another thing to resolve is whether it can reference `e1` itself, which can be argued either way.

**Comment by jimmyfrasche** [comment]
On a page with only `input` and `label` (and no other elements using a `for` attribute), I assume that `label /idref(for)/ input` and `[for] /idref(for)/ *` and `* /idref(for)/ *` would all match the same elements. Are the `*` necessary or would it suffice to write just `/idref(for)/`?

Would the more specific `/for/` also match inputs wrapped in a label?

**Comment by keithamus** [comment]
Perhaps worth noting that the HTML spec makes no mention of "idref", nor does MDN. HTML refers to the reflection of an element from an attribute as an ["attr-associated element"](https://html.spec.whatwg.org/#attr-associated-element), and MDN doesn't really strongly associate these IDL properties or content attributes with a specifically named concept. While "idref" might be used as a name by some developers, it seems to be an archaic term that is perhaps borrowed from XSLT, which I believe defined it?

The point being, while I hold no objection to using idref (others might though), I think there is lots of opportunity for us to pick a better name that is either more ergonomic or more meaningful for developers.

**Comment by LeaVerou** [comment]
> Perhaps worth noting that the HTML spec makes no mention of "idref", nor does MDN. HTML refers to the reflection of an element from an attribute as an ["attr-associated element"](https://html.spec.whatwg.org/#attr-associated-element), and MDN doesn't really strongly associate these IDL properties or content attributes with a specifically named concept. While "idref" might be used as a name by some developers, it seems to be an archaic term that is perhaps borrowed from XSLT, which I believe defined it?
> 
> The point being, while I hold no objection to using idref (others might though), I think there is lots of opportunity for us to pick a better name that is either more ergonomic or more meaningful for developers.

Agreed (archaic indeed — the name comes from XML), though I can't think of a better name for the concept that is hardcoded to ids. Perhaps that's another argument for a parameterized version where both the source attribute and the referencing attribute are passed as parameters every time (see #12446). If we call it `ref()` it is only 1 character longer than `idref()`.
